File: CPP/Algorithm/Shallowest_Decomposition_Tree.cpp
#define log __lg
#define ctz __builtin_ctz

// Rooted tree
struct arborescence {
    vector<vector<int64_t>> children;
    int64_t rt;
};

arborescence sdt(vector<vector<int64_t>> &gp, auto rt = 0) {
    int n = gp.size();
    vector<vector<int64_t>> decomposition_tree(n), stacks(log(n) + 1);
    auto extract_chain = [&](int64_t labels, int64_t u) {
        while (labels) {
            int label = log(labels); labels ^= 1ll << label;
            int64_t v = stacks[label].back(); stacks[label].pop_back();
            decomposition_tree[u].push_back(v); u = v;
        }
    };
    vector<int64_t> forbidden(n, -1);
    auto dfs = [&](int64_t u, int64_t p, auto&& self) -> void {
        int64_t forbidden_once = 0, forbidden_twice = 0;
        for (auto &v : gp[u]) {
            if (v != p) {
                self(v, u, self);
                forbidden_twice |= forbidden_once & (forbidden[v] + 1);
                forbidden_once |= forbidden[v] + 1;
            }
        }
        forbidden[u] = forbidden_once | ((1ll << log((forbidden_twice << 1)  + 1)) - 1);
        int64_t label_u = ctz(forbidden[u] + 1);
        stacks[label_u].push_back(u);
        for (int i = gp[u].size() - 1; i >= 0; --i) {
            extract_chain((forbidden[gp[u][i]] + 1) & ((1ll << label_u) - 1), u);
        }
    };
    dfs(rt, -1, dfs);
    int64_t max_label = log(forbidden[rt] + 1), drt = stacks[max_label].back();
    stacks[max_label].pop_back();
    extract_chain((forbidden[rt] + 1) & ((1ll << max_label) - 1), drt);
    return {move(decomposition_tree), drt};
}

File: Python/Shallowest_Decomposition_Tree.py
def ctz(x): return (x & -x).bit_length() - 1
def shallowest_decomposition_tree(graph, root=0):
    n = len(graph); forbidden = [0] * n
    decomposition_tree = [[] for _ in range(n)]
    stacks = [[] for _ in range(n.bit_length())]
    def extract_chain(labels, u):
        while labels:
            label = labels.bit_length() - 1; labels ^= 2**label
            v = stacks[label].pop(); decomposition_tree[u].append(v); u = v
    dfs = [root]
    while dfs:
        u = dfs.pop()
        if u >= 0:
            forbidden[u] = -1; dfs.append(~u)
            for v in graph[u]:
                if not forbidden[v]: dfs.append(v)
        else:
            u = ~u; forbidden_once = forbidden_twice = 0
            for v in graph[u]:
                forbidden_twice |= forbidden_once & (forbidden[v] + 1)
                forbidden_once |= forbidden[v] + 1
            forbidden[u] = forbidden_once | ((1 << forbidden_twice.bit_length()) - 1)
            label_u = ctz(forbidden[u] + 1); stacks[label_u].append(u)
            for v in graph[u]:
                extract_chain((forbidden[v] + 1) & ((1 << label_u) - 1), u)
    max_label = (forbidden[root] + 1).bit_length() - 1
    decomposition_root = stacks[max_label].pop()
    extract_chain((forbidden[root] + 1) & ((1 << max_label) - 1), decomposition_root)
    return decomposition_tree, decomposition_root